engine/dlib/src/test/test_httpclient.cpp

```diff
@@ -848,6 +848,53 @@ TEST_P(dmHttpClientTest, MaxAgeCache)
     ASSERT_EQ(dmHttpCache::RESULT_OK, cache_r);
 }
 
+TEST_P(dmHttpClientTest, TrustCache)
+{
+    dmHttpClient::Delete(m_Client);
+
+    // Reinit client with http-cache
+    dmHttpClient::NewParams params;
+    params.m_Userdata = this;
+    params.m_HttpContent = dmHttpClientTest::HttpContent;
+    params.m_HttpHeader = dmHttpClientTest::HttpHeader;
+    params.m_DNSChannel = m_DNSChannel;
+    params.m_TrustCache = true;
+    dmHttpCache::NewParams cache_params;
+    cache_params.m_Path = "tmp/cache";
+    dmHttpCache::Result cache_r = dmHttpCache::Open(&cache_params, &params.m_HttpCache);
+    ASSERT_EQ(dmHttpCache::RESULT_OK, cache_r);
+    m_Client = dmHttpClient::New(&params, m_URI.m_Hostname, m_URI.m_Port, strcmp(m_URI.m_Scheme, "https") == 0);
+    ASSERT_NE((void*) 0, m_Client);
+
+    dmHttpClient::Result r;
+    r = dmHttpClient::Get(m_Client, "/max-age-cached");
+    ASSERT_TRUE((r == dmHttpClient::RESULT_OK && m_StatusCode == 200) || (r == dmHttpClient::RESULT_NOT_200_OK && m_StatusCode == 304));
+    std::string val = m_Content;
+
+    // The web-server is returning System.currentTimeMillis()
+    // Ensure next ms
+    dmTime::Sleep(1U);
+
+    m_Content = "";
+    r = dmHttpClient::Get(m_Client, "/max-age-cached");
+    ASSERT_TRUE((r == dmHttpClient::RESULT_OK && m_StatusCode == 200) || (r == dmHttpClient::RESULT_NOT_200_OK && m_StatusCode == 304));
+    ASSERT_STREQ(m_Content.c_str(), val.c_str());
+
+    // max-age is 1 second
+    dmTime::Sleep(1000000U);
+
+    // The cached response has expired since it is older than max-age but
+    // we have told the client to trust the cache and use a cached response
+    // if it exists
+    m_Content = "";
+    r = dmHttpClient::Get(m_Client, "/max-age-cached");
+    ASSERT_TRUE((r == dmHttpClient::RESULT_OK && m_StatusCode == 200) || (r == dmHttpClient::RESULT_NOT_200_OK && m_StatusCode == 304));
```

Comment:
> "But for some type of resources we want to use local cache without real outside call (like MaxAge header). Bc of cache policy we make call and receive 304 error code each time."

How are we currently handling the max age?
It seems it should be like you've implemented it, but without the need for an extra "trust" policy?

Reply:
Is the problem that we didn't get a max-age header from the server?

Reply:
It is my understanding that the cache policy defined for an instance of dmHttpCache affects the entire cache so we can't use that to solve this particular problem.

The current implementation if there is a previous response with an ETag is to check the ETag against the server after max-age time has elapsed (the trust policy for the http cache is NOT CONSISTENCY_POLICY_TRUST_CACHE). I'm not sure if all servers return a max-age header with an ETag. I don't think it is required, although it would make sense if they did return a max-age with the ETag.

In any case, the problem we are trying to solve is to provide developers with a way to tell our http client implementation to use a cached result regardless if it has expired (older than max-age) and regardless if it has an ETag.

Reply:
I think a problem is that we don't fully understand the client's use case?

I interpret that quote one way, and you seem to have a different understanding of it.

I interpret it as that they might not _know_ we have support for max-age-header?

> ..Bc of cache policy we make call and receive 304 error code each time."

This suggests that something strange is going on. Doesn't the server have support for max-age-header?
Can it be turned on?

> In any case, the problem we are trying to solve is to provide developers with a way to tell our http client implementation to use a cached result regardless if it has expired (older than max-age) and regardless if it has an ETag.

True, we also try to avoid adding things that might not be needed.

Reply:
> "We use Google Cloud CDN and we can’t configure headers on their side. CDN always returns Etag header."

What headers did _you_ get when testing the url they gave you?
Didn't it return a max-age-header?

It seems Google Cloud CDN supports max-age:
https://cloud.google.com/cdn/docs/caching

Reply:
Another issue with the new trust policy, is that the user now have to implement something to know when to re-request the data from the server.

An idea could be that the http request call accepts a "max-age" override, for those cases when the server doesn't supply the header.

Reply:
The response contains the following headers:

```
    alt-svc = "clear",
    expires = "Tue, 09 Mar 2021 08:23:05 GMT",
    date = "Mon, 08 Mar 2021 08:23:05 GMT",
    age = "87",
    etag = ""0cbea4495f0e6bda6048b6b1a6227793"",
    cache-control = "public, max-age=86400"
```

BUT what I'm noticing now is that we prioritise the Etag over max-age header when storing the response in the cache:

https://github.com/defold/defold/blob/dev/engine/dlib/src/dlib/http_client.cpp#L961-L968

What this means is that the client will always check the Etag since m_Valid will never be true (max-age is 0 since we prioritise the Etag).

Also the Etag will not be ok since the http cache doesn't have a trust policy set to CONSISTENCY_POLICY_TRUST_CACHE and we also need to do at least one roundtrip to get a 304 and set the cached resource m_Verified to true.

https://github.com/defold/defold/blob/dev/engine/dlib/src/dlib/http_client.cpp#L1095-L1096

It would make more sense to set the Etag AND the max-age and only check the Etag if max-age time has elapsed. But I'm concerned that this will have implications we have not foreseen.